src/dask_awkward/io/parquet.py

```diff
@@ -0,0 +1,519 @@
+                    form_key="",
+                )
+            else:
+                return forms.ByteMaskedForm(
+                    "i8",
+                    content,
+                    valid_when=True,
+                    form_key="",
+                )
+        else:
+            return content
+
+    def contains_record(form):
+        if isinstance(form, forms.RecordForm):
+            return True
+        elif isinstance(form, forms.ListOffsetForm):
+            return contains_record(form.content)
+        else:
+            return False
+
+    def recurse(arrow_type, path):
+        if isinstance(arrow_type, pyarrow.StructType):
+            names = []
+            contents = []
+            for index in range(arrow_type.num_fields):
+                field = arrow_type[index]
+                names.append(field.name)
+                content = maybe_nullable(
+                    field, recurse(field.type, path + (field.name,))
+                )
+                contents.append(content)
+            assert len(contents) != 0
+            return forms.RecordForm(contents, names)
+
+        elif isinstance(arrow_type, pyarrow.ListType):
+            field = arrow_type.value_field
+            content = maybe_nullable(
+                field, recurse(field.type, path + ("list", "item"))
+            )
+            return forms.ListOffsetForm("i32", content, form_key="")
+
+        elif isinstance(arrow_type, pyarrow.LargeListType):
+            field = arrow_type.value_field
+            content = maybe_nullable(
+                field, recurse(field.type, path + ("list", "item"))
+            )
+            return forms.ListOffsetForm("i64", content, form_key="")
+
+        elif arrow_type == pyarrow.string():
+            return forms.ListOffsetForm(
+                "i32",
+                forms.NumpyForm("uint8"),
+                parameters={"__array__": "string"},
+                form_key="",
+            )
+
+        elif arrow_type == pyarrow.large_string():
+            return forms.ListOffsetForm(
+                "i64",
+                forms.NumpyForm("uint8"),
+                parameters={"__array__": "string"},
+                form_key="",
+            )
+
+        elif arrow_type == pyarrow.binary():
+            return forms.ListOffsetForm(
+                "i32",
+                forms.NumpyForm("uint8"),
+                parameters={"__array__": "bytestring"},
+                form_key="",
+            )
+
+        elif arrow_type == pyarrow.large_binary():
+            return forms.ListOffsetForm(
+                "i64",
+                forms.NumpyForm("uint8"),
+                parameters={"__array__": "bytestring"},
+                form_key="",
+            )
+
+        elif isinstance(arrow_type, pyarrow.DataType):
+            if arrow_type == pyarrow.null():
+                return forms.EmptyForm(form_key="")
+            else:
+                dtype = np.dtype(arrow_type.to_pandas_dtype())
```

Comment:
@martindurant when trying to load the chicago taxi dataset I'm getting an exception here of the form:
```
> /Users/ddavis/software/repos/dask-awkward/src/dask_awkward/io/parquet.py(161)recurse()
-> dtype = np.dtype(arrow_type.to_pandas_dtype())
(Pdb) np.dtype(arrow_type.to_pandas_dtype())
*** NotImplementedError: dictionary<values=string, indices=int32, ordered=0>
(Pdb) arrow_type
DictionaryType(dictionary<values=string, indices=int32, ordered=0>)
```
where the `arrow_type` can't use `to_pandas_dtype()`.

Reply:
That is clearly *not* a numpy type, but it is something that awkward can handle (indexed array?).

`_parquet_scheme_to_form` was written some time ago by @jpivarski . We should probably switch to using ak.metadata_from_parquet , after checking that it copes with this input.

Reply:
Arrow DictionaryType is equivalent to Awkward IndexedArray. `ak._v2.from_arrow` makes that translation and puts the word "categorical" in the IndexedArray's parameters; `ak._v2.to_arrow` only makes that translation if it finds the word "categorical". The reason is that if you build an Arrow Dictionary with non-unique values, it segfaults, and in general, Awkward IndexedArrays don't forbid non-unique values. The "categorical" parameter is a promise that they're unique.

The alternative to preserving a DictionaryType or IndexedArray is to evaluate it (`project` in IndexedArray). The data in this structure is a lazy array-slice:

```python
dictionary_values[dictionary_index]
```

where both are arrays. So if the Arrow Dictionary's values are a NumPy dtype, applying this procedure would give you a NumPy dtype. The type of the Dictionary is the type of its values.

Isn't `ak.metadata_from_parquet` just a high-level wrapper of `_parquet_scheme_to_form`? When the dust settles, we should have just one implementation of Arrow Schema ↔ Awkward Form. (I think that's already the case: I don't remember seeing two implementations of it.)

Reply:
> Isn't ak.metadata_from_parquet just a high-level wrapper of _parquet_scheme_to_form

No, it uses `ak._v2._connect.pyarrow.form_handle_arrow`. `_parquet_schema_to_form` only exists in dask-awkward.